### PR TITLE
Patch LangChain loading functions to handle pickle serialization issue

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/11582?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11582/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11582
```

</p>
</details>

### What changes are proposed in this pull request?

Since `langchain-community` >= 0.0.27, saving or loading a module that relies on pickle serialization requires `allow_dangerous_deserialization` flag set to `True`. This broke the model loading in the LangChain flavor.

This was not solvable on MLflow side until [this PR](https://github.com/langchain-ai/langchain/pull/18894)was merged to LangChain on `0.1.14`. Before this patch, the `allow_dangerous_deserialization` argument cannot be passed via higher-level loading function like `load_chain()`, `load_llm()`, which we are using inside the LangChain flavor.

**Note**: The patching logic can be improved to be consistent with how we've done in #11644, but it requires good amount of refactoring. To avoid complicating the diff and risking the release schedule, this PR implements the minimum change to make the feature work, and refactoring is done in a separate draft PR here: #11677 

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?
TODO
- [ ] No. You can skip the rest of this section. 
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/langchain/utils.py [hunk 1]
```diff
@@ -422,6 +436,56 @@ def _get_path_by_key(root_path, key, conf):
...
+        return loader_func
+
+    import langchain
+
+    if Version(langchain.__version__) >= Version("0.1.14"):
+        # For LangChain 0.1.14 and above, we can pass `allow_dangerous_deserialization` flag
+        # via the loader APIs. Since the model is serialized by the user (or someone who has
+        # access to the tracking server), it is safe to set this flag to True.
+        def patched_loader(*args, **kwargs):
+            return loader_func(*args, **kwargs, allow_dangerous_deserialization=True)
```
Comment (harupy) [inline mlflow/langchain/utils.py:467]: Any chance that `kwargs` already has `allow_dangerous_deserialization`?

Comment (harupy) [reply mlflow/langchain/utils.py:467]: It looks like there is no chance for `kwargs` to include `allow_dangerous_deserialization`.

File: mlflow/langchain/utils.py [hunk 2]
```diff
@@ -422,6 +436,56 @@ def _get_path_by_key(root_path, key, conf):
...
+            try:
+                return loader_func(*args, **kwargs)
+            except ValueError as e:
+                if "This code relies on the pickle module" in str(e):
+                    raise MlflowException(
+                        "Since langchain-community 0.0.27, loading a module that relies on "
+                        "the pickle deserialization requires the `allow_dangerous_deserialization` "
+                        "flag to be set to True when loading. However, this flag is not supported "
+                        "by the installed version of LangChain. Please upgrade LangChain to 0.1.14 "
+                        "or above by running `pip install langchain>=0.1.14 -U`.",
```
Comment (harupy) [inline mlflow/langchain/utils.py:480]:
```suggestion
                        "or above by running `pip install langchain>=0.1.14`.",
```

I think `-U` can be removed. If 0.1.14 is already installed, no update is required.

File: tests/langchain/test_langchain_model_export.py [hunk 3]
```diff
@@ -2244,6 +2244,46 @@ def test_pyfunc_builtin_chat_response_conversion_fails_gracefully():
...
+    mlflow.langchain.save_model(chain, model_path)
+
+    if IS_PICKLE_SERIALIZATION_RESTRICTED and Version(langchain.__version__) < Version("0.1.14"):
+        # For LangChain between 0.1.12 and 0.1.14, MLflow cannot load a model that relies on pickle
+        # serialization, instead, raises an MlflowException with a message that explains the issue.
+        with pytest.raises(MlflowException, match=r"Since langchain-community 0.0.27, loading a"):
+            mlflow.langchain.load_model(model_path)
+        return
+    else:
+        loaded_model = mlflow.langchain.load_model(model_path)
```
Comment (harupy) [inline tests/langchain/test_langchain_model_export.py:2279]:
```suggestion
    loaded_model = mlflow.langchain.load_model(model_path)
```

nit: can we remove else here?

File: mlflow/langchain/utils.py (hunk 1 above)
Comment (B-Step62) [reply mlflow/langchain/utils.py:467]: Yup!

File: mlflow/langchain/utils.py (hunk 2 above)
Comment (B-Step62) [reply mlflow/langchain/utils.py:480]: That's right, we don't need it as we specify version bound.

File: tests/langchain/test_langchain_model_export.py [hunk 4]
```diff
@@ -833,6 +829,10 @@ def load_db(persist_dir):
     version.parse(langchain.__version__) < version.parse("0.0.297"),
     reason="Saving SQLDatabaseChain chains requires langchain>=0.0.297",
 )
+@pytest.mark.skipif(
+    version.parse(langchain.__version__) in (version.parse("0.1.14"), version.parse("0.1.15")),
```
Comment (serena-ruan) [inline tests/langchain/test_langchain_model_export.py:833]: Is dev version affected?

Comment (serena-ruan) [reply tests/langchain/test_langchain_model_export.py:833]: oh looks like dev tests pass

**Review by serena-ruan — APPROVED** [review]
LGTM!

File: tests/langchain/test_langchain_model_export.py (hunk 4 above)
Comment (B-Step62) [reply tests/langchain/test_langchain_model_export.py:833]: Yes! the fix is just merged yesterday🙂

File: tests/langchain/test_langchain_model_export.py [hunk 5]
```diff
@@ -833,6 +829,10 @@ def load_db(persist_dir):
...
+    reason="LangChain 0.1.14 and 0.1.15 has a bug in loading SQLDatabaseChain",
```
Comment (harupy) [inline tests/langchain/test_langchain_model_export.py:834]:
```suggestion
    reason="LangChain 0.1.14 and 0.1.15 have a bug in loading SQLDatabaseChain",
```

nit

**Review by harupy — APPROVED** [review]
LGTM!

**Review by BenWilson2 — APPROVED** [review]
LGTM! Let's ship

**Pull request merged**